x-pack/plugins/index_management/public/application/components/mappings_editor/lib/utils.ts

```diff
@@ -602,3 +603,93 @@ export const deNormalizeRuntimeFields = (fields: NormalizedRuntimeFields): Runti
     };
   }, {} as RuntimeFields);
 };
+
+/**
+ * get all the fields from given state which matches selected DataTypes from filter
+ *
+ * @param state The state that we are using depending on the context (when adding new fields, static state is used)
+ * @param filteredDataTypes data types array from which fields are filtered from given state
+ */
+
+export const getFieldsMatchingFilterFromState = (
+  state: State,
+  filteredDataTypes: string[]
+): {
+  [id: string]: NormalizedField;
+} => {
+  const getFieldId = (fieldId: string) => {
+    if (filteredDataTypes.includes(TYPE_DEFINITION[state.fields.byId[fieldId].source.type].label)) {
+      return fieldId;
+    } else {
+      return undefined;
+    }
+  };
+
+  const getfieldIds = Object.entries(state.fields.byId).map(([key, _]) => getFieldId(key));
+
+  return Object.fromEntries(
+    Object.entries(state.fields.byId).filter(([id]) => getfieldIds.includes(id))
+  );
+};
+
+/** accepts Generics argument and returns value, if value is not null or undefined
+ * @param value
+ */
+function isNotNullish<T>(value: T | null | undefined): value is T {
+  return value !== null && value !== undefined;
+}
+
+/** returns normalized field that matches the dataTypes from the filteredDataTypes array
+ * @param normalizedFields fields that we are using, depending on the context (when adding new fields, static state is used)
+ * @param filteredDataTypes data types array from which fields are filtered from given state. When there are no filter selected, array would be undefined
+ */
+export const getFieldsFromState = (
+  normalizedFields: NormalizedFields,
+  filteredDataTypes?: string[]
+): NormalizedField[] => {
+  const getField = (fieldId: string) => {
+    if (filteredDataTypes) {
+      if (
+        filteredDataTypes.includes(
+          TYPE_DEFINITION[normalizedFields.byId[fieldId].source.type].label
+        )
+      ) {
+        return normalizedFields.byId[fieldId];
+      }
+    } else {
+      return normalizedFields.byId[fieldId];
+    }
+  };
+  const fields: Array<NormalizedField | undefined> = filteredDataTypes
+    ? Object.entries(normalizedFields.byId).map(([key, _]) => getField(key))
+    : normalizedFields.rootLevelFields.map((id) => getField(id));
+  return fields.filter(isNotNullish);
+};
+/**
+ * returns true if given value is first occurence of array
+ * useful when filtering unique values of an array
+ */
+function filterUnique<T>(value: T, index: number, array: T[]) {
+  return array.indexOf(value) === index;
+}
+/**
+ * returns array consisting of all field types from state's fields including nested fields
+ * @param fields
+ */
+const getallFieldsIncludingNestedFields = (fields: Fields) => {
+  const fieldArray: DataType[] = [];
+  const fieldsValue = Object.values(fields);
+  for (const field of fieldsValue) {
+    if (field.type) fieldArray.push(field.type);
+    if (field.fields) getallFieldsIncludingNestedFields(field.fields);
```

Comment:
We don't do anything with the results of the nested fields in the array in this line or the next. So we compute them and then just throw them away. I assume instead we want to add the contents of the resulting arrays to the current fieldArray.

We should also be adding unit tests for these functions so we can catch these issues.

Reply:
`getallFieldsIncludingNestedFields` should have `fieldArray` as argument. Essentially, this functions fetches all field types including nested fields as well. 

I missed this completely. Thanks for highlighting this 🙌

Reply:
> We should also be adding unit tests for these functions so we can catch these issues.
Agreed! https://github.com/elastic/kibana/pull/181188/commits/f7c5232178674c317658caaf71c371181e8db7e3